networkit/cpp/centrality/DynApproxBetweenness.cpp

```diff
@@ -1,178 +1,268 @@
+        count source = sources[sorted[i]];
+        // forward BFS
+        marked1[source] = true;
+        count dist1 = 0;
+        q.push(source);
+        do {
+            node u = q.front();
+            q.pop();
+            G.forNeighborsOf(u, [&](node v) {
+                if (!marked1[v] && scc.componentOfNode(u) == scc.componentOfNode(v)) {
+                    qNext.push(v);
+                    marked1[v] = true;
+                }
+            });
+            if (q.empty() && !qNext.empty()) {
+                q.swap(qNext);
+                ++dist1;
+            }
+        } while (!q.empty());
+
+        // backward BFS
+        marked2[source] = true;
+        count dist2 = 0;
+        q.push(source);
+        do {
+            node u = q.front();
+            q.pop();
+            G.forInNeighborsOf(u, [&](node v) {
+                if (!marked2[v] && scc.componentOfNode(u) == scc.componentOfNode(v)) {
+                    qNext.push(v);
+                    marked2[v] = true;
+                }
+            });
+            if (q.empty() && !qNext.empty()) {
+                q.swap(qNext);
+                ++dist2;
+            }
+        } while (!q.empty());
+        // source node is not taken into account in any of the two approximations
+        vdapproxs[sorted[i]] = dist1 + dist2 + 1;
+        if (vdapproxs[sorted[i]] == 0) // for singletons
+            vdapproxs[sorted[i]] = 1;
+    }
+    // starting from the bottom, we compute all the cumulative vd approximations
+    count vd = vdapproxs[0];
+    std::vector<count> vdSuccessors(sccNum, 0);
+    for (count i = 0; i <= sccNum - 1; i++) {
+        node c = sorted[sccNum - i];
+        if (sccDAG.degreeOut(c) > 0) {
+            vdapproxs[c] += vdSuccessors[c];
+        }
+        if (vdapproxs[c] > vd) {
+            vd = vdapproxs[c];
+        }
+        sccDAG.forInEdgesOf(c, [&](node c, node c_pred) {
+            if (vdSuccessors[c_pred] < vdapproxs[c])
+                vdSuccessors[c_pred] = vdapproxs[c];
+        });
+    }
+    return vd;
+}
 
 count DynApproxBetweenness::getNumberOfSamples() {
     return r;
 }
 
-
 void DynApproxBetweenness::run() {
-    if (G.isDirected()) {
-        throw std::runtime_error("Invalid argument: G must be undirected.");
-    }
     scoreData.clear();
     scoreData.resize(G.upperNodeIdBound());
     u.clear();
     v.clear();
     sampledPaths.clear();
-
-    Diameter diam(G, DiameterAlgo::estimatedPedantic);
-    diam.run();
-    edgeweight vd = diam.getDiameter().first;
-
-    INFO("estimated diameter: ", vd);
-    r = ceil((universalConstant / (epsilon * epsilon)) * (floor(log2(vd - 2)) + 1 - log(delta)));
-    INFO("taking ", r, " path samples");
+    count vd;
+    if (G.isDirected()) {
+        vd = computeVDdirected();
+    } else {
+        Diameter diam(G, DiameterAlgo::estimatedPedantic);
+        diam.run();
+        vd = diam.getDiameter().first;
+    }
+    r = ceil((universalConstant / (epsilon * epsilon))
```

Comment:
Math functions such as `ceil`, `floor` etc. are defined in the deprecated `math.h` C library. I recommend using their counterparts defined in the `cmath` library (i.e., `std::ceil`, `std::floor`, etc).